Add troubleshooting page for SignalR (#20125)
dotnet/AspNetCore.Docs: aspnetcore/signalr/troubleshoot.md

---
title: ASP.NET Core SignalR connection troubleshooting
author: 
description: ASP.NET Core SignalR connection troubleshooting.
monikerRange: '>= aspnetcore-2.1'
ms.author: 
ms.custom: mvc
ms.date: 04/08/2020
no-loc: ["ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: signalr/troubleshoot
---
# Troubleshoot connection errors

This section provides help with errors that can occur when trying to establish a connection to a ASP.NET Core SignalR hub.

### Response code 404

When using WebSockets and `skipNegotiation = true`
```log
WebSocket connection to 'wss://xxx/HubName' failed: Error during WebSocket handshake: Unexpected response code: 404
```

> * If the connection takes too long to send a request to the server after the negotiate the server will delete the ID and return a 404 if the connection uses that ID

Reviewer: 👍

> * If the connection takes too long to send a request to the server after the negotiate the server will delete the ID and return a 404 if the connection uses that ID

Reviewer 2: This is difficult for native speakers to read and impossible to MT (machine translate)

```suggestion
* If the connection takes too long to send a request to the server after the negotiate, the server:

  * Deletes the ID.
  * Returns a 404 if the connection uses that ID.
```

Reviewer 2: Does it only return 404 if the connection uses that ID? Maybe the following is better:
```suggestion
* If the connection uses the ID and takes too long to send a request to the server after the negotiate, the server:

  * Deletes the ID.
  * Returns a 404.
```

> * If you have multiple servers and aren't using sticky sessions, the connection can start on one server and then switch to another server that doesn't know about it

Reviewer 2: ```suggestion
* When using multiple servers without sticky sessions, the connection can start on one server and then switch to another server. The other server is not aware of the previous connection.
```

> * Double check that the client is connecting to the correct endpoint. e.g. Server is hosted at http://127.0.0.1:5000/hub/myHub and client is trying to connect to http://127.0.0.1:5000/myHub.

Reviewer 2: ```suggestion
* Verify the client is connecting to the correct endpoint. For example, the server is hosted at `http://127.0.0.1:5000/hub/myHub` and client is trying to connect to `http://127.0.0.1:5000/myHub`.
```

### Response code 400 or 503

For the following error:

```log
WebSocket connection to 'wss://xxx/HubName' failed: Error during WebSocket handshake: Unexpected response code: 400

Error: Failed to start the connection: Error: There was an error with the transport.
```

> The error is usually caused by client only use WebSockets transport but WebSockets protocol is not enabled on server.

Reviewer 2: ```suggestion
This error is usually caused by client using only WebSockets transport but the WebSockets protocol is not enabled on the server.
```

### Response code 307

When using WebSockets and `skipNegotiation = true`
```log
WebSocket connection to 'ws://xxx/HubName' failed: Error during WebSocket handshake: Unexpected response code: 307
```

> This can also happen during the negotiate request.

Reviewer 2: ```suggestion
This error can also happen during the negotiate request.
```

---
title: ASP.NET Core SignalR connection troubleshooting
author: 
description: ASP.NET Core SignalR connection troubleshooting.
monikerRange: '>= aspnetcore-2.1'

> ms.author: 

Reviewer 3: @Rick-Anderson I see your screenshot with the suggested pending changes to add @bradygaster as an author to this doc. Did you mean to submit a new review with those suggestion?

Reviewer: same question.

ms.custom: mvc
ms.date: 04/08/2020
no-loc: ["ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: signalr/troubleshoot
---
# Troubleshoot connection errors

This section provides help with errors that can occur when trying to establish a connection to a ASP.NET Core SignalR hub.

### Response code 404

When using WebSockets and `skipNegotiation = true`
```log
WebSocket connection to 'wss://xxx/HubName' failed: Error during WebSocket handshake: Unexpected response code: 404
```

* If the connection uses the ID and takes too long to send a request to the server after the negotiate, the server:

  * Deletes the ID.

>   * Returns a 404.

Reviewer 3: This is probably the least common cause of a 404, don't you think? I would probably order the bullet points with bad path first, no sticky session second and this last.

* When using multiple servers without sticky sessions, the connection can start on one server and then switch to another server. The other server is not aware of the previous connection.
* Verify the client is connecting to the correct endpoint. For example, the server is hosted at `http://127.0.0.1:5000/hub/myHub` and client is trying to connect to `http://127.0.0.1:5000/myHub`.

### Response code 400 or 503

For the following error:

```log
WebSocket connection to 'wss://xxx/HubName' failed: Error during WebSocket handshake: Unexpected response code: 400

Error: Failed to start the connection: Error: There was an error with the transport.
```

This error is usually caused by client using only WebSockets transport but the WebSockets protocol is not enabled on the server.

### Response code 307

When using WebSockets and `skipNegotiation = true`
```log
WebSocket connection to 'ws://xxx/HubName' failed: Error during WebSocket handshake: Unexpected response code: 307
```

> This error can also happen during the negotiate request.

Reviewer 3: Do any of the HTTP clients and/or WebSocket clients SignalR uses not follow 307 redirects? I just searched the aspnetcore repo for "SignalR 307" (without quotes) and came up with no reports of an issue like that.

Reviewer 3: Found it: https://stackoverflow.com/questions/63437471/websocket-connection-to-ws-localhost5000-notificationhub-failed-error-duri. I'm still not sure that this could "happen during the negotiate request." Do you have evidence of that?

Author: Yes, I point out one possible cause below

Reviewer 3: I don't see it. Why would a 307 break the negotiate request itself?

Author: Because we don't follow redirects

Reviewer 3: Oh. Why? Because we assume it will break WebSockets?

Author: Hmm, I'm not sure. I thought it was because auth handlers often redirect instead of returning 401, but we set a header to try and avoid that...

Author: https://github.com/dotnet/aspnetcore/issues/27068
